Support for GlobusComputeExecutor

yadudoc opened the pull request:
# Description

This PR adds a new `GlobusComputeExecutor` that wraps the Globus Compute SDK to allow Parsl to execute tasks via Globus Compute. This mechanism supports remote execution of tasks similar to the functionality that `parsl.channels` enabled and is a potential replacement.

Since `GlobusCompute` often runs on remote machines that do not have a shared-filesystem with the parsl runtime, tests have been updated with a new `shared_fs` and `staging_required` pytest markers. I have not added tests and CI actions to enable executing these tests against our CI system, but you can run tests locally with these steps:

1. Install `globus-compute-sdk` with `pip install .[globus-compute]`
2. Configure and start a globus-compute-endpoint, `globus-compute-endpoint start <endpoint_name>`
3. Set an env var with the endpoint id for tests: `export GLOBUS_COMPUTE_ENDPOINT=<endpoint_id>`
4. Run tests with `pytest -v -k "not shared_fs" --config parsl/tests/configs/globus_compute.py parsl/tests/`



# Changed Behaviour

N/A

# Fixes

Fixes # (issue)

## Type of change

Choose which options apply, and delete the ones which do not apply.

- New feature

benclifford commented:
Aside from a lack of CI test, my main concern (from seeing people implement this before) is what goes on with different Parsl versions: there's one version in the endpoint (pinned to a specific version by Globus Compute) and another version on the submit side, and having different Parsl versions like that is out of scope for Parsl.

This might be a case of documenting what users should expect to work or not work, or might be something deeper. At the very least we should be expecting them to be able to use the same combination of versions as used in CI.

rjmello commented on parsl/tests/conftest.py:
```diff
@@ -163,6 +163,14 @@ def pytest_configure(config):
         'markers',
         'executor_supports_std_stream_tuples: Marks tests that require tuple support for stdout/stderr'
     )
+    config.addinivalue_line(
+        'markers',
+        'globus_compute: Marks tests that require a valid globus_compute target'
+    )
```
Do any tests use this marker?

yadudoc replied:
Unused. I've removed this marker.

rjmello commented on parsl/tests/configs/globus_compute.py:
```diff
@@ -0,0 +1,18 @@
+
+def fresh_config():
+
+    endpoint_id = os.environ.get("GLOBUS_COMPUTE_ENDPOINT",
+                                 "4b116d3c-1703-4f8f-9f6f-39921e5864df")
```
While anyone could potentially run tests against our tutorial EP, I don't think we should suggest doing so.

yadudoc replied:
Good call. I've changed this to require GLOBUS_COMPUTE_ENDPOINT to be specified for tests. Failing with a KeyError should be clear enough that something is misconfigured.

rjmello commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,139 @@
+    for more details.
+    """
+
+    def __init__(
+            self,
+            endpoint_id: Optional[UUID_LIKE_T] = None,
+            task_group_id: Optional[UUID_LIKE_T] = None,
+            resource_specification: Optional[dict[str, Any]] = None,
+            user_endpoint_config: Optional[dict[str, Any]] = None,
+            label: str = "GlobusComputeExecutor",
+            batch_size: int = 128,
+            amqp_port: Optional[int] = None,
+            **kwargs,
+    ):
```
Consider adding the `client` parameter

yadudoc replied:
I'm hesitant to add this because it will require `globus_compute_sdk.sdk.Client` type hints when this is an Optional module. Right now, passing this in via **kwargs will still work fine. What do you think?

rjmello replied:
I think we should add it to clearly communicate support for the parameter. I've already had an early adopter ask about this.

To add type hints, we could do something like this:
```python
try:
    from globus_compute_sdk import Client
except ImportError:
    Client = Any
 
 class GlobusComputeExecutor(ParslExecutor, RepresentationMixin):
    def __init__(
            self,
            endpoint_id: Optional[UUID_LIKE_T] = None,
            client: Optional[Client] = None,
            ...
```

or

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from globus_compute_sdk import Client

 class GlobusComputeExecutor(ParslExecutor, RepresentationMixin):
    def __init__(
            self,
            endpoint_id: Optional[UUID_LIKE_T] = None,
            client: Optional["Client"] = None,
            ...
```

rjmello replied:
haha thanks Ben :metal:

yadudoc replied:
Fixed now.

rjmello replied:
I could be missing something, but I don't see the `client` parameter.

yadudoc replied:
It's here -> https://github.com/Parsl/parsl/pull/3607/files#diff-03350429687012cb1a3ed39ac9d3549f2b4c825555f7e9dee0600ff82b16164aR48

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,139 @@
+
+        Future
+        """
+        self._executor.resource_specification = resource_specification or self.resource_specification
```
if you want to specify that per task, it maybe makes sense to make it a new entry in the task submission resource_specification - that's the mechanism for specify everything about "how" a task is to be executed in Parsl

christopherwharrop-noaa replied:
Unless I'm misunderstanding something, that is how the multi-user endpoints are meant to work.  If you can't specify it at submit time, then the alternative will be to have multiple instantiations of `GlobusComputeExecutor` for the same endpoint id, but with different `user_endpoint_config`.  Maybe that's not a "bad thing", but while you would reduce number of endpoints to maintain, Parsl workflows would still have to have one `GlobusComputeExecutor` for each type of endpoint you want to interact with, even when on the same machine.  And maybe that's not so bad because even if you didn't have to do that, you'd still need to handle the multiple 'user_endpoint_config` dicts that are required for the various endpoint configs you need.  My thought was to match what is available on the Globus Compute side, which allows (I think) setting that at both Executor instantiation time and at submit time.

benclifford replied:
i think per-task specification is a good option to make available. i was commenting on *where* that specification should go when submitting from Parsl.

christopherwharrop-noaa replied:
Would that look something like this?:
```
self._executor.user_endpoint_config = resource_specification["user_endpoint_config"] or self.user_endpoint_config
```

benclifford replied:
yes

yadudoc replied:
I've made some changes to allow `resource_specification` and `user_endpoint_config` to both work.

```
# Specifying MPI apps
py_app(resource_specification={'num_nodes': 2, 'ranks_per_node':4})

# Specifying user_endpoint_config
py_app(resource_specification={'user_endpoint_config': {"account_id": "user_allocation_account_id"}}

# Specifying both:
py_app(resource_specification={'num_nodes': 2, 'ranks_per_node':4,
'user_endpoint_config': {"account_id": "user_allocation_account_id"}
})

```

benclifford commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,112 @@
+        source .venv/bin/activate
+        make deps
+        make clean_coverage
+        # Installing parsl into venv required for GCendpoint
```
doesn't GC dependency handling bring in Parsl?

yadudoc replied:
Correct. This is now fixed.

christopherwharrop-noaa commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        self._executor.resource_specification = resource_specification or self.resource_specification
+        # Pop user_endpoint_config since it is illegal in resource_spec for globus_compute
+        self._executor.user_endpoint_config = resource_specification.pop('user_endpoint_config', self.user_endpoint_config)
```
Should the `pop` happen before you assign `resource_specification` to `self._executor.resource_specification`?

benclifford replied:
I think: the `pop` live-edits the value previously assigned to self._executor.resource_specification - because it is assignment of mutable objects, not of values.

But following that same reasoning, @yadudoc does this mean this code is editing the resource_specification dict provided by the user? (at some arbitrary point in the future after the user invokes an app...?)

christopherwharrop-noaa replied:
Oh, right!  Of course. I'm mixing up my languages.

yadudoc replied:
I've updated this section to pop `user_endpoint_config` from a copy of `resource_specification` to avoid modifying the user-supplied dict.

benclifford commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,110 @@
+        python3 -c "import globus_compute_endpoint; print(globus_compute_endpoint.__version__)"
+        cat << EOF > /home/runner/.globus_compute/default/config.yaml
+        engine:
+          type: ThreadPoolEngine
```
I feel like this is a little evasive... the thread pool engine isn't what Globus Compute is "really" intended for, right?

yadudoc replied:
Right, this was added for 2 reasons. 1) The default engine had issues spawning all the components in the same virtualenv, despite worker_init being configured. 2) Even with a single worker, tests were slow which I suspect is due to the endpoint being way more resource-intensive. 

From what I see, what I'm testing is the executor on the parsl side. Parsl tests needn't care about what engine the GC endpoint uses on its side.

benclifford commented on docs/userguide/execution.rst:
```diff
@@ -89,6 +89,9 @@ Parsl currently supports the following executors:
 These executors cover a broad range of execution requirements. As with other Parsl components, there is a standard interface (ParslExecutor) that can be implemented to add support for other executors.
 
+5. `parsl.executors.globus_compute.GlobusComputeExecutor`: This executor uses `Globus Compute <https://globus-compute.readthedocs.io/en/latest/index.html>`_
+as the execution backend to run functions on remote systems.
```
most of the other paragraphs use the word "task" here, not "function"

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,162 @@
+
+        label:
+            a label to name the executor; mainly utilized for
+            logging and advanced needs with multiple executors.
```
"advanced needs" is weird phrasing that sounds like bluffing about ... something?

yadudoc replied:
Sounds odd now that you pointed it out. Fixed.

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,162 @@
+        )
+
+    def start(self) -> None:
+        """Empty function
```
probably don't need this docstring at all. or justify the why here if there's something interesting to talk about.

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,162 @@
+        pass
+
+    def submit(self, func: Callable, resource_specification: Dict[str, Any], *args: Any, **kwargs: Any) -> Future:
+        """ Submit fn to globus-compute
```
Use `func` not `fn`

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,162 @@
+        self._executor.user_endpoint_config = resource_specification.pop('user_endpoint_config', self.user_endpoint_config)
+        return self._executor.submit(func, *args, **kwargs)
+
+    def shutdown(self, wait=True, *, cancel_futures=False):
```
what are all these parameters for?

yadudoc replied:
Removed.

benclifford commented on parsl/tests/test_bash_apps/test_basic.py:
```diff
@@ -24,6 +24,7 @@ def foo(x, y, z=10, stdout=None, label=None):
     return f"echo {x} {y} {z}"
 
 
+@pytest.mark.shared_fs
```
maybe all these shared_fs / staging_required markers should be merged as a separate PR? They're not specific to this test case (eg. they came I think originally from the kubernetes testing draft PR #3482 and that work would be slightly advanced by these being merged to master, independent of the rest of this PR being merged or not)

yadudoc replied:
Fair enough. I left this here because those changes wouldn't kick in without tests that exclude based on the marker. I've split out these changes over on this PR -> https://github.com/Parsl/parsl/pull/3669

yadudoc commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,110 @@
+      - opened
+      - synchronize
+
+  workflow_dispatch:
```
Obsolete. It was used to trigger the workflow run manually for debugging.

benclifford commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,103 @@
+        make clean_coverage
+        
+        # Temporary fix, until changes make it into compute releases
+        git clone -b configure_tasks_working_dir https://github.com/globus/globus-compute.git        
```
maybe its good to wait for Globus Compute to support the necessary features in a release, so that this can pin a relevant supported release?

benclifford replied:
(does this mean that right now a user shouldn't expect to be able to use an actual Globus Compute but instead needs to install this fork?)

yadudoc replied:
The github action uses `ThreadPoolEngine` because it is lightweight compared to running the default config that uses `GlobusComputeEngine`. The branch that I'm pointing to fixes a `ThreadPoolEngine` only bug. You can run these same tests against a local GlobusCompute endpoint, by setting `GLOBUS_COMPUTE_ENDPOINT=<ep_id>` and I've done those tests against a few configs. I believe Chris (NOAA) has also done a bunch of testing at this point.

christopherwharrop-noaa replied:
I have been testing the `GlobusComputeExecutor` via this branch in combination with `globus-compute-endpoint==2.30.1`.  I am using that release, not a special branch/tag.  The one feature I haven't yet tested is passing in the `user_endpoint_config` at call time rather than at `GlobusComputeExecutor` instantiation time.  What I've done so far is working great!

benclifford commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,103 @@
+        export GLOBUS_COMPUTE_ENDPOINT=$(globus-compute-endpoint list | grep default | cut -c 3-38)      
+        echo "GLOBUS_COMPUTE_ENDPOINT = $GLOBUS_COMPUTE_ENDPOINT"
+
+        # temporary; until test-matrixification
```
nothing so temporary as boilerplate copied from someone else's unfinished business.

benclifford commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,103 @@
+        name: runinfo-${{ matrix.python-version }}-${{ steps.job-info.outputs.as-ascii }}-${{ github.sha }}
+        path: |
+          runinfo/
+          .pytest/
```
its not called .pytest any more - see PR #3631

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,166 @@
+        ----------
+
+        endpoint_id:
+            id of the endpoint to which to submit tasks
```
what happens when this is `None`?

yadudoc replied:
I'll add an assertion to force this. There's a discrepancy between the executor and the `compute_sdk.Executor` in that an endpoint_id is required at init since we don't have a mechanism in Parsl to update the `endpoint_id` dynamically at runtime.

yadudoc replied:
I've added typeguard to `__init__` to cover this.

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,166 @@
+
+        kwargs:
+            Other kwargs listed will be passed through to globus_compute_sdk.Executor
+            as is
```
link to where I can find out what those are?

yadudoc replied:
Done

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,166 @@
+        )
+
+    def start(self) -> None:
+        pass
```
In the Parsl configuration model, this separation of init and start means that the code will try to connect to an executor at the point that a user defines a configuration that uses globus compute, rather than at the point that they initialize Parsl with a Globus compute.  I'm not clear on what this Executor does at the point that it is instantiated (for example, making network connections?).

That might be out of line with other code which doesn't do anything until it is `start`ed - for example, when is the interchange started in a HighThroughputExecutor configuration?

yadudoc replied:
I've moved the actual `globus_compute_sdk.Executor` instantiation to start.

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,166 @@
+
+        self._executor.resource_specification = res_spec
+        self._executor.user_endpoint_config = user_endpoint_config
+        return self._executor.submit(func, *args, **kwargs)
```
This feels a bit horribly thread unsafe, were Parsl ever to get multithreaded submission - the sort of thing DFK.submitter_lock was put in place to deal with.

yadudoc replied:
Potentially yes. I don't see how this is only a GCE specific problem. Secondly, I think it's safe to wait until users report this as an issue, or ask for it before working on this.

benclifford replied:
When this is used from the DFK, this is already in a lock (although as you're both well aware, there is strong community pressure to use pieces of Parsl outside of the DFK). That lock was introduced in PR #625 because executors are not expected to be thread safe on submission, and so in that context this code is not dangerous.

This is a more general backpressure against using this style of API that seems to conflate the state of the submission system as a whole with parameters to a single task execution - I've definitely fixed concurrency bugs in Parsl because of this coding style before, that lead not to Parsl errors but to subtle misexecutions that mostly look plausible.

benclifford replied:
(see #3492, #1413)

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,166 @@
+        GCE.shutdown will cancel all futures that have not yet registered with
+        Globus Compute and will not wait for the launched futures to complete.
+        """
+        return self._executor.shutdown(wait=False, cancel_futures=True)
```
shutdown returns a `None`, not whatever this returns (which is hopefully also None)

yadudoc replied:
It returns `None`.

yadudoc commented:
@benclifford @khk-globus Thanks for the review and all the comments on this issue. I believe I've addressed all of them and at this point, we are only waiting for a release of `globus-compute-endpoint` with fixes from [ComputePR #1689](https://github.com/globus/globus-compute/pull/1689) to update the CI testing target.

khk-globus commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,171 @@
+
+        GCE.shutdown will cancel all futures that have not yet registered with
+        Globus Compute and will not wait for the launched futures to complete.
+        """
+        self._executor.shutdown(wait=False, cancel_futures=True)
```
If a user (or Parsl developer) mistakenly calls `.shutdown()` on a GCE instance that hasn't been started, this currently explodes.  What about being more friendly with a simple check:

```python
if not self._executor:
    raise RuntimeError("Executor not started.")
```

That's just a smidgen less cryptic than `None has no attribute .shutdown`

The same suggestion goes for `.submit()`

yadudoc replied:
Fixed now as `self.executor` is set at init.

khk-globus commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,171 @@
+        self.client = client
+        self.executor_kwargs = kwargs
+
+        if not _globus_compute_enabled:
+            raise OptionalModuleMissing(
+                ['globus-compute-sdk'],
+                "GlobusComputeExecutor requires globus-compute-sdk installed"
+            )
```
To elide useless work in the `_globus_compute_enabled is False` case, consider making this the very first statement in `__init__()`.  No need to do the the other work if it's going to be thrown away anyway.  (I get that this change is minimal, but it's the principal of fail early and fail fast that I'm after.)

khk-globus reviewed (commented):
I haven't finished reading through the PR, but these are my first thoughts.  I'll chime in on Monday morning (-0400).

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,171 @@
+        kwargs:
+            Other kwargs listed will be passed through to globus_compute_sdk.Executor
+            as is. Refer to `globus-compute docs
+            <https://globus-compute.readthedocs.io/en/latest/reference/executor.html#globus-compute-executor>`_
+        """
+        super().__init__()
+        self.endpoint_id = endpoint_id
+        self.task_group_id = task_group_id
```
There's some part of me that looks at this and says " 'just' pass in a Globus Compute `executor` as the single parameter" and let users read the Globus Compute docs to understand the object they are creating. I think right now, users of this code are going to have to be understanding stuff about Globus Compute anyway.

I think it gives a better "lazy path" sustainability story.

yadudoc replied:
@benclifford I like the idea of passing in the `globus_compute_sdk.Executor` itself to the executor, but this would break the `init -> start` separation that `ParslExecutors` follow.

yadudoc replied:
We just discussed this, and the consensus is that passing along the executor is the most straightforward solution here.

benclifford commented on parsl/executors/globus_compute.py:
```diff
@@ -0,0 +1,171 @@
+            client: Optional[Client] = None,
+            **kwargs,
+    ):
+        """
+        Parameters
+        ----------
+
+        endpoint_id:
```
more fuel for https://github.com/Parsl/parsl/pull/3607#discussion_r1844962325

khk-globus commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,97 @@
+        make deps
+        make clean_coverage
+
+        # Temporary fix until fixes make it to a release
+        git clone -b main https://github.com/globus/globus-compute.git
+        pip3 install globus-compute/compute_sdk globus-compute/compute_endpoint
```
Is the `git clone` still necessary after the next GC release, come this Thursday?

yadudoc replied:
No, this is a stopgap until the release.

khk-globus commented on setup.py:
```diff
@@ -45,6 +45,7 @@
     'flux': ['pyyaml', 'cffi', 'jsonschema'],
     'proxystore': ['proxystore'],
     'radical-pilot': ['radical.pilot==1.90', 'radical.utils==1.90'],
+    'globus_compute': ['globus_compute_sdk>=2.27.1'],
     # Disabling psi-j since github direct links are not allowed by pypi
```
It's been a minute since we started this PR; might increase this to the most up-to-date as of ~now (2.34.0)

yadudoc replied:
Fixed.

khk-globus commented on .github/workflows/gce_test.yaml:
```diff
@@ -0,0 +1,97 @@
+        which globus-compute-endpoint
+        python3 -c "import globus_compute_sdk; print(globus_compute_sdk.__version__)"
+        python3 -c "import globus_compute_endpoint; print(globus_compute_endpoint.__version__)"
+        cat << EOF > /home/runner/.globus_compute/default/config.yaml
+        engine:
+          type: ThreadPoolEngine
+          max_workers: 4
+        EOF
+        cat /home/runner/.globus_compute/default/config.yaml
+        mkdir ~/.globus_compute/default/tasks_working_dir
+        globus-compute-endpoint start default
+        globus-compute-endpoint list
```
Discussed out-of-band.  This is not feasible to do under the possibility of concurrently running these tests.

yadudoc commented:
@khk-globus I've added some configuration examples and squashed all the previous changes. 

I've opted to keep the `executor.shutdown` mechanism using `result_watcher.shutdown`. I agree that this isn't the best solution here, but the other approach we discussed was causing hangs. Canceling the futures to have the `result_watcher` exit when the exit handler is triggered appears to cause a hang. From the logs, it looks like the atexit handler triggered RW.shutdown, but instead of recognizing that the futures were canceled, it does a blocking wait until the task completes, which then throws an error since the future is already canceled. This smells like a bug. I've not figured out the root cause on the `globus_compute_sdk.Executor` side, but I do not think we should hold this PR up for that.
